File: src/main/java/com/techtitans/smartbudget/api/model/CompanyData.java
package com.techtitans.smartbudget.api.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.PastOrPresent;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@NoArgsConstructor
@AllArgsConstructor
@Data
@Table(name = "company_data", schema = "public")
public class CompanyData {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id_data")
    private long id_data;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_company", referencedColumnName = "id_company")
    @NotNull(message = "Company cannot be null")
    private Companies company;

    @Column(name = "open")
    @DecimalMin(value = "0.0", inclusive = false, message = "Open must be greater than 0")
    private double open;

    @Column(name = "high")
    @DecimalMin(value = "0.0", inclusive = false, message = "High must be greater than 0")
    private double high;

    @Column(name = "low")
    @DecimalMin(value = "0.0", inclusive = false, message = "Low must be greater than 0")
    private double low;

    @Column(name = "close")
    @DecimalMin(value = "0.0", inclusive = false, message = "Close must be greater than 0")
    private double close;

    @Column(name = "date")
    @Temporal(TemporalType.TIMESTAMP)
    @PastOrPresent(message = "Date must be in the past or present")
    private LocalDateTime date;

}

File: AI_RFS_Test/models/company_data.py
class CompanyData:
    def __init__(self, id, company, open, high, low, close, date):
        self.id = id
        self.company = company
        self.open = open
        self.high = high
        self.low = low
        self.close = close
        self.date = date

    def __str__(self):
        return f'{self.company} on {self.date} with open {self.open} and close {self.close}'

    def __repr__(self):
        return f'{self.company} on {self.date}'

    def __eq__(self, other):
        return self.company == other.company and self.date == other.date

    def __hash__(self):
        return hash((self.company, self.date))
